Program capture sharp bits page for v0.41 (#7062)
PennyLaneAI/pennylane: doc/introduction/program_capture_sharp_bits.rst

Our goal with program-capture is to support *more* than just the core features of the 
PennyLane you have come to know and love, but there are some **quirks and restrictions 
to be aware of while we strive towards that ideal**. In this document, we provide 
an overview of said constraints.

.. note::
    #. From here onwards, ``qml.capture.enable`` is assumed to be present within 
    the scope of code examples, unless otherwise stated. This ensures that program-capture
    is enabled.

    #. Using program-capture requires that JAX be installed. Please consult the 
    JAX documentation for `installation instructions <https://docs.jax.dev/en/latest/installation.html>`__.
    
    #. Our short name for PennyLane code that is captured as jaxpr is *plxpr* (PennyLane-jaxpr).
    Program-capture and plxpr can be considered synonymous and interchangeable. 

.. _device_compatibility:

Device compatibility 
--------------------

Currently, ``default.qubit`` and ``lightning.qubit`` are compatible with program-capture.
In addition, ``default.qubit`` requires that ``wires`` be specified. This is in 
contrast to when program-capture is disabled, where automatic qubit management takes

> place internally.

Reviewer: Doesn't "lightning.qubit" need `wires` to be specified too?

```python
qml.capture.enable()

dev = qml.device("lightning.qubit")

@qml.qnode(dev, autograph=True)
def circuit():
    qml.X(0)
    return qml.state()

>>> circuit()
NotImplementedError: devices must specify wires for integration with plxpr capture.
```

Author: Correct! I can mention both, as I'm assuming that the user (1) knows that DQ can be used without wires when capture is disabled and (2) knows that wires has always been required in LQ. Lots of assumptions there :)

.. role:: html(raw)
   :format: html

> Program-capture sharp bits and debugging tips
> =============================================

Reviewer 2: ```suggestion
Program-capture sharp bits
==========================
```
Just a suggestion, to avoid this taking up two lines in the sidebar

> Program-capture is PennyLane's new internal representation for hyrbid quantum-classical 
> programs that maintains the same familiar feeling frontend that you're used to, 
> while providing better performance, harmonious integration with just-in-time compilation 
> frameworks like `Catalyst <https://docs.pennylane.ai/projects/catalyst/en/stable/index.html>`__ 
> (:func:`~.pennylane.qjit`) and JAX-jit, and compact representations of programs that preserve 
> their structure.

Reviewer 2: I really like this intro, but it doesn't read quite right to me 🤔 Program capture is the _feature_, while `plxpr` is the intermediate representation. Maybe this?

```suggestion
Program capture is a new feature of PennyLane, that allows PennyLane
to capture and understand more details about the hybrid workflow --
including quantum operations, classical processing, structured control flow,
and dynamicism.

This allows for better performance, and more functional and dynamic
workflows, and better integration with just-in-time compilation frameworks
such as
`Catalyst <https://docs.pennylane.ai/projects/catalyst>`__ 
(via the :func:`~.pennylane.qjit` decorator) and JAX.

Internally, this is supported by representing the program via a new
intermediate representation (`plxpr`), rather than a quantum tape.
```

Reviewer 2: I also made the text a bit more dry to match our other reference docs 😅

Program-capture in PennyLane is propped up by JAX's internal representation of programs 
called `jaxpr <https://docs.jax.dev/en/latest/jaxpr.html>`__, which works by leveraging 
the Python interpreter to extract the core elements of programs and capture them 
into a light-weight language.

In part because of PennyLane's NumPy-like syntax and functionality, quantum programs 
written with PennyLane translate nicely into the language of jaxpr, letting your 
quantum-classical programs burn the same fuel that powers JAX and just-in-time compilation.

The quantum tape/script has been PennyLane's trusty program representation since 
its inception, but our vision with program-capture is to supplant the quantum tape
as the default program representation in PennyLane, and to support *more* than just 

> the core features of the PennyLane you have come to know and love. 

Reviewer 2: If the main purpose of this guide is usability, this is internal technical detail that is not needed

So I suggest removing it

There are some **quirks and restrictions to be aware of while we strive towards 
that ideal**. Additionally, we've added backwards-compatibility features that make 
the transition from tape-based code to program-capture be a smooth one. In this 
document, we provide an overview of the constraints, "gotchas" to be aware of, and

> features that will help get your existing tape-based code working with program-capture.

Reviewer 2: nice!

.. note::

    #. From here onwards, :func:`~.pennylane.capture.enable` is assumed to be present 
    within the scope of code examples, unless otherwise stated. This ensures that 

>     program-capture is enabled.

Reviewer 2: For reference documentation, since it is more likely to be consulted than read end-to-end, I wouldn't actually mind repeating enable in the codeblocks!

It means we'll have examples that when copied and pasted into fresh notebooks will work as shown

> Program-capture in PennyLane is propped up by JAX's internal representation of programs 
> called `jaxpr <https://docs.jax.dev/en/latest/jaxpr.html>`__, which works by leveraging 
> the Python interpreter to extract the core elements of programs and capture them 
> into a light-weight language.
> 
> In part because of PennyLane's NumPy-like syntax and functionality, quantum programs 
> written with PennyLane translate nicely into the language of jaxpr, letting your 
> quantum-classical programs burn the same fuel that powers JAX and just-in-time compilation.
> 
> The quantum tape/script has been PennyLane's trusty program representation since 
> its inception, but our vision with program-capture is to supplant the quantum tape
> as the default program representation in PennyLane, and to support *more* than just 
> the core features of the PennyLane you have come to know and love. 
> 
> There are some **quirks and restrictions to be aware of while we strive towards 
> that ideal**. Additionally, we've added backwards-compatibility features that make 
> the transition from tape-based code to program-capture be a smooth one. In this 
> document, we provide an overview of the constraints, "gotchas" to be aware of, and
> features that will help get your existing tape-based code working with program-capture.
> 
> .. note::
> 
>     #. From here onwards, :func:`~.pennylane.capture.enable` is assumed to be present 
>     within the scope of code examples, unless otherwise stated. This ensures that 
>     program-capture is enabled.
> 
>     #. Using program-capture requires that JAX be installed. Please consult the 
>     JAX documentation for `installation instructions <https://docs.jax.dev/en/latest/installation.html>`__.
>     
>     #. Our short name for PennyLane code that is captured as jaxpr is *plxpr* (PennyLane-jaxpr).
>     "Program-capture" and "plxpr" can be considered synonymous and interchangeable. 

Author: ```suggestion
Program-capture is a new feature of PennyLane that allows for compactly expressing 
details about hybrid workflows, including quantum operations, classical processing, 
structured control flow, and dynamicism.

This new feature unlocks better performance, more functional and dynamic workflows, 
and smoother integration with just-in-time compilation frameworks like 
`Catalyst <https://docs.pennylane.ai/projects/catalyst/en/stable/index.html>`__ 
(via the :func:`~.pennylane.qjit` decorator) and JAX-jit.

Internally, program-capture is supported by representing hyrbid programs via a new 
intermediate representation (IR) called ``plxpr``, rather than a quantum tape. The 
``plxpr`` IR is an adaptation of JAX's intermediate representation of programs (``jaxpr``).

Our vision with ``plxpr`` is to supplant the quantum tape as our default program 
representation, and to support *more* than just the core features of PennyLane.

There are some **quirks and restrictions to be aware of while we strive towards 
that ideal**. Additionally, we've added backwards-compatibility features that make 
the transition from tape-based code to program-capture be a smooth one. In this 
document, we provide an overview of the constraints, "gotchas" to be aware of, and
features that will help get your existing tape-based code working with program-capture.

.. note::

    #. From here onwards, :func:`~.pennylane.capture.enable` is assumed to be present 
    within the scope of code examples, unless otherwise stated. This ensures that 
    program-capture is enabled.

    #. Using program-capture requires that JAX be installed. Please consult the 
    JAX documentation for `installation instructions <https://docs.jax.dev/en/latest/installation.html>`__.
```

@josh146 I reworked the intro to mush together your comments with some of the parts I have in there and how they flowed together.

Let me know what you think!

I

Reviewer 2: :+1:

>     "Program-capture" and "plxpr" can be considered synonymous and interchangeable. 

Reviewer 2: Mentioned in another comment that they don't seem synonymous to me -- one is the feature you are enabling, and the other is simply the IR.

Device wires 
------------

> Currently, ``default.qubit`` and ``lightning.qubit`` are compatible with program-capture.

Reviewer 2: This sounds like a sharp bit, that not all devices might be compatible with program capture!

> With program-capture enabled, both ``lightning.qubit`` and ``default.qubit`` require 

Reviewer 2: should it be 'program capture', without the hyphen? it reads a bit odd to me with the caption

Author: If "program capture" is the feature, shouldn't it be treated as one word?

Reviewer 2: not necessarily, features can be more than one word! E.g., Qualtran integration is a feature.

Reviewer 2: and parameter broadcasting

Author: Idk to me it feels natural to hyphenate all of those, but I suppose that's not the standard we've adopted so I'll change it 😅

that ``wires`` be specified at device instantiation (this is in contrast to when 
program-capture is disabled, where automatic qubit management takes place internally
with ``default.qubit``).

.. code-block:: python

    @qml.qnode(qml.device('default.qubit'))
    def circuit():
        qml.Hadamard(0)
        return qml.state()

>>> circuit()

> NotImplementedError: devices must specify wires for integration with plxpr capture.

Reviewer 2: Probably this should say 'program capture', not plxpr capture?

Author: I have a PR for this :) https://github.com/PennyLaneAI/pennylane/pull/7130

Author: ```suggestion
NotImplementedError: devices must specify wires for integration with program capture.
```

Valid JAX data types 
--------------------

Because of the nature of creating and executing plxpr, it is **best practice to 
use JAX-compatible types whenever possible**, in particular for arguments to quantum 
functions and QNodes, and positional arguments in PennyLane gate operations. JAX-compatible 

> types are ``jax.numpy`` arrays and standard Python ``int``\ s and ``float``\ s. 

Reviewer 2: I'm a bit confused, aren't JAX-compatible types anything that is pytree compatible? So numpy arrays, dictionaries, etc?

Reviewer 2: This is what we have in the [Catalyst sharp bits guide](https://docs.pennylane.ai/projects/catalyst/en/latest/dev/sharp_bits.html#function-argument-restrictions):

> Compiled functions can accept arbitrary function arguments, as long as the inputs can be represented as [Pytrees](https://jax.readthedocs.io/en/latest/pytrees.html) — tree-like structures built out of Python container objects such as lists, dictionaries, and tuples — where the values (leaf nodes) are compatible types.
>
> Compatible types includes Booleans, Python numeric types, JAX arrays, and PennyLane quantum operators.
> 
> Note: Non-numeric types, such as strings, are generally not supported as arguments to compiled functions.

Author: ```suggestion
functions and QNodes, and positional arguments in PennyLane gate operations. Examples of 
JAX-compatible types are ``jax.numpy`` arrays, regular NumPy arrays, dictionaries, standard
Python ``int``\ s and ``float``\ s, and anything else with a valid `Pytree <https://jax.readthedocs.io/en/latest/pytrees.html>` 
representation.
```

For example, ``list``\ s, ``range``\ s, and strings are not valid JAX types for 
the positional argument in :class:`~.pennylane.MultiRZ`, and will result in an error:

> .. code-block:: python
>     dev = qml.device('default.qubit', wires=2)

Reviewer 2: ```suggestion
.. code-block:: python

    dev = qml.device('default.qubit', wires=2)
```

    @qml.qnode(dev)
    def circuit():
        qml.MultiRZ([0.1, 0.2], wires=[0, 1])
        return qml.expval(qml.X(0))

>>> circuit()
TypeError: Value [0.1, 0.2] with type <class 'list'> is not a valid JAX type

Providing a ``list`` as input to a quantum function or QNode is accepted in cases 
where the ``list`` is being indexed into, thereby retrieving a valid JAX type:

.. code-block:: python

    dev = qml.device('default.qubit', wires=2)

    @qml.qnode(dev)
    def circuit(x):
        qml.RZ(x[0], wires=0)
        qml.RX(x[1], wires=1)
        return qml.expval(qml.X(0))

>>> circuit([0.1, 0.2])
Array(0., dtype=float32)

JAX-incompatible types, like Python ``range``\ s, are acceptable as keyword arguments:

.. code-block:: python
    
    dev = qml.device('default.qubit', wires=2)

    @qml.qnode(dev)
    def circuit(x, range_of_wires=None):
        for w in range_of_wires:
            qml.RZ(x[0], wires=w)
            qml.RX(x[1], wires=w)

        return qml.expval(qml.X(0))

>>> circuit([0.1, 0.2], range_of_wires=range(2))
Array(0., dtype=float32)

But, again, using JAX-compatible types wherever possible is recommended.

Positional arguments

> ====================

Reviewer 2: (incorrect heading type)

Positional arguments **must** be acceptable JAX types. For instance, consider this
example with ``qml.RZ``:

.. code-block:: python

    import jax.numpy as jnp

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circui(angle):
        qml.RX(phi=angle, wires=0)
        return qml.expval(qml.Z(0))

>>> angle = jnp.array(0.1)

> >>> circuit1(angle)

Reviewer 2: I'm not sure I understand -- this is a valid jax type?

Author: Typo :)

...
UnexpectedTracerError: Encountered an unexpected tracer. A function transformed by JAX had a side effect, allowing for a reference to an intermediate value with type float32[] wrapped in a DynamicJaxprTracer to escape the scope of the transformation.
...

But, when the angle is passed as a positional argument, the circuit executes as 
expected:

.. code-block:: python

    @qml.qnode(dev)
    def circui(angle):
        qml.RX(angle, wires=0)
        return qml.expval(qml.Z(0))

>>> angle = jnp.array(0.1)
>>> circuit1(angle)
Array(0.9950042, dtype=float32)

> Parameter broadcasting and ``vmap``
> -----------------------------------

Reviewer 2: ```suggestion
Parameter broadcasting and vmap
-------------------------------
```

just so the docs render a bit better

> Parameter-broadcasting is generally not compatible with program-capture. There are 

Reviewer 2: do we use hyphens in other docs when we refer to parameter broadcasting?

Author: Not sure 🤔

cases that magically work, but one shouldn't extrapolate beyond those particular 
cases.

Instead, it is best practice to `use jax.vmap <https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html>`__:

.. code-block:: python

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit(x):
        qml.RX(x, wires=0)
        return qml.expval(qml.Z(0))

>>> x = jnp.array([0.1, 0.2, 0.3])
>>> vmap_circuit = jax.vmap(circuit)
>>> vmap_circuit(x)
Array([0.9950042 , 0.9800666 , 0.95533645], dtype=float32)

More information for using ``jax.vmap`` can be found in the 
`JAX documentation <https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html#jax.vmap>`__.

Transforms
----------

One of the core features of PennyLane is modularity, which has allowed users to 
transform QNodes in a NumPy-like way and to create their own transforms with ease. 
Your favourite transforms will still work with program-capture enabled (including

> custom transforms), but **decorating QNodes with just ``@transform_name`` will not 

Reviewer 2: ReST doesn't support nested formatting

all :class:`~.pennylane.RX`` gates to the end of a circuit.

.. code-block:: python

    @qml.transform
    def shift_rx_to_end(tape):
        """Transform that moves all RX gates to the end of the operations list."""
        new_ops, rxs = [], []

        for op in tape.operations:
            if isinstance(op, qml.RX):
                rxs.append(op)
            else:
                new_ops.append(op)
        
        operations = new_ops + rxs
        new_tape = tape.copy(operations=operations)
        return [new_tape], lambda res: res[0]

Decorating with just ``@shift_rx_to_end`` will not work, and will give a vague error:

.. code-block:: python

    @shift_rx_to_end
    @qml.qnode(qml.device("default.qubit", wires=1))
    def circuit():
        qml.RX(0.1, wires=0)
        qml.H(wires=0)
        return qml.state()

>>> print(qml.draw(circuit)())
...
NotImplementedError: 

A requirement for tape transforms to be compatible with program capture is to further 
decorate QNodes with the experimental :func:`~.pennylane.capture.expand_plxpr_transforms` 
decorator:

.. code-block:: python

    @qml.capture.expand_plxpr_transforms
    @shift_rx_to_end

>     @qml.qnode(qml.device("default.qubit", wires=1))

Reviewer 2: Is this something we will be able to automate?

    def circuit():
        qml.RX(0.1, wires=0)
        qml.H(wires=0)
        return qml.state()

>>> print(qml.draw(circuit)())
0: ──H──RX(0.10)─┤  State

Dynamic variables and transforms

> ================================

Reviewer 2: (incorrect heading type)

> Some transforms in the :module:`~.pennylane.transforms` API have bespoke plxpr implementations
> that *directly* transform jaxpr:

Reviewer 2: This is a bit confusing (leans into technical details) -- better to state the main take away for users. E.g.,

```suggestion
Some transforms in the :doc:`/code/qml_transforms` module have natively support program capture mode:
```

#. :func:`~.pennylane.transforms.merge_rotations`
#. :func:`~.pennylane.transforms.single_qubit_fusion`
#. :func:`~.pennylane.transforms.unitary_to_rot`
#. :func:`~.pennylane.transforms.merge_amplitude_embedding`
#. :func:`~.pennylane.transforms.commute_controlled`
#. :func:`~.pennylane.transforms.decompose`
#. :func:`~.pennylane.map_wires`
#. :func:`~.pennylane.transforms.cancel_inverses`

> **This section applies to any transform not listed above, including custom quantum tape
> transforms**.
> 
> There are a few other restrictions on QNodes and applying transforms that do not 
> have a native program-capture implementation:

Reviewer 2: I find this section a bit hard to parse, because of this qualifier. You could simply say

```suggestion
For transforms that do not natively work with program capture, they can continue to be used with certain limitations:
```

For reference docs, always pays to be as concise as possible :) People will be skim reading, and trying to extract information quickly

> #. Transforms that return multiple tapes are not backwards-compatible.
> #. Transforms that return non-trivial post-processing functions are not backwards-compatible.

Reviewer 2: Best to be explicit here; rather than saying backwards compatible, best to say 
```suggestion
#. Transforms that return multiple tapes are not supported.
#. Transforms that return non-trivial post-processing functions are not supported.
```

#. Tape transforms will fail to execute if the transformed quantum function or QNode contains:

>     #. ``qml.cond`` with dynamic parameters as predicates.
>     #. ``qml.for_loop`` with dynamic parameters for ``start``, ``stop``, or ``step``.
>     #. ``qml.while_loop``.

Reviewer 2: ```suggestion

   #. ``qml.cond`` with dynamic parameters as predicates.
   #. ``qml.for_loop`` with dynamic parameters for ``start``, ``stop``, or ``step``.
   #. ``qml.while_loop``.
```

Here is an example with our toy ``shift_rx_to_end`` transform and a dynamic parameter
for ``stop`` in ``qml.for_loop``.

.. code-block:: python

    @qml.capture.expand_plxpr_transforms
    @shift_rx_to_end
    @qml.qnode(qml.device("default.qubit", wires=4))
    def circuit(stop):

        @qml.for_loop(0, stop, 1)
        def loop(i):
            qml.RX(0.1, wires=i)
            qml.H(wires=i)
        
        loop(stop)

        return qml.state()

>>> circuit(4)
TracerIntegerConversionError: The __index__() method was called on traced array with shape int32[].
The error occurred while tracing the function wrapper at /Users/isaac/.virtualenvs/pl-latest/lib/python3.11/site-packages/pennylane/transforms/core/transform_dispatcher.py:41 for make_jaxpr. This concrete value was not available in Python because it depends on the value of the argument inner_args[0].
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerIntegerConversionError

> ``while`` loops 
> ---------------

Reviewer 2: ```suggestion
while loops 
-----------
```

function:

.. code-block:: python

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit():

        @qml.while_loop(lambda a: a > 3)
        def loop(a):
            a += 1
            return a

        a = 0
        loop(a)

        qml.RX(0, wires=0)
        return qml.state()

>>> circuit()
...
KeyError: <gast.gast.Lambda object at 0x136ff82b0>

As a workaround, use a regular Python function:

.. code-block:: python

    dev = qml.device("default.qubit", wires=1)

    def func(x):
        return x > 3

    @qml.qnode(dev)
    def circuit():

        @qml.while_loop(func)
        def loop(a):
            a += 1
            return a

        a = 0
        loop(a)
        
        qml.RX(0, wires=0)
        return qml.state()

>>> circuit()
Array([1.+0.j, 0.+0.j], dtype=complex64)

Calculating operator matrices in QNodes
---------------------------------------

The matrix of an operator cannot be computed with :func:`~.pennylane.matirx` within
a QNode, and will raise an error:

.. code-block:: python

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit():

>         mat = qml.matrix(qml.X(0))

Reviewer 2: does this work if used as `qml.matrix(qml.X)(0)`?

Author: No but good point! It raises a different error.

>>> angle = jnp.array(0.1)
>>> circuit1(angle)
Array(0.9950042, dtype=float32)

Parameter broadcasting and vmap
-------------------------------

Parameter broadcasting is generally not compatible with program-capture. There are 
cases that magically work, but one shouldn't extrapolate beyond those particular 
cases.

Instead, it is best practice to `use jax.vmap <https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html>`__:

.. code-block:: python

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit(x):
        qml.RX(x, wires=0)
        return qml.expval(qml.Z(0))

>>> x = jnp.array([0.1, 0.2, 0.3])
>>> vmap_circuit = jax.vmap(circuit)
>>> vmap_circuit(x)
Array([0.9950042 , 0.9800666 , 0.95533645], dtype=float32)

More information for using ``jax.vmap`` can be found in the 
`JAX documentation <https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html#jax.vmap>`__.

Transforms
----------

One of the core features of PennyLane is modularity, which has allowed users to 
transform QNodes in a NumPy-like way and to create their own transforms with ease. 
Your favourite transforms will still work with program-capture enabled (including

> custom transforms), but **decorating QNodes with just ``@transform_name`` will not 
> work** and will give a vague error. Additionally, decorating QNodes with the experimental 

Author: ```suggestion
custom transforms), but decorating QNodes with just ``@transform_name`` **will not 
work** and will give a vague error. Additionally, decorating QNodes with the experimental 
```

.. role:: html(raw)
   :format: html

> Program-capture sharp bits

Author: TODO: add a section on using with qjit (experimental capture = True flag)

==========================

Program-capture is PennyLane's new internal representation for hyrbid quantum-classical 
programs that maintains the same familiar feeling frontend that you're used to, 
while providing better performance, harmonious integration with just-in-time compilation 
frameworks like `Catalyst <https://docs.pennylane.ai/projects/catalyst/en/stable/index.html>`__ 
(:func:`~.pennylane.qjit`) and JAX-jit, and compact representations of programs that preserve 
their structure.

Program-capture in PennyLane is propped up by JAX's internal representation of programs 
called `jaxpr <https://docs.jax.dev/en/latest/jaxpr.html>`__, which works by leveraging 
the Python interpreter to extract the core elements of programs and capture them 

> into a light-weight language.

Reviewer: ```suggestion
into a light-weight representation.
```

In part because of PennyLane's NumPy-like syntax and functionality, quantum programs 
written with PennyLane translate nicely into the language of jaxpr, letting your 
quantum-classical programs burn the same fuel that powers JAX and just-in-time compilation.

The quantum tape/script has been PennyLane's trusty program representation since 
its inception, but our vision with program-capture is to supplant the quantum tape
as the default program representation in PennyLane, and to support *more* than just 
the core features of the PennyLane you have come to know and love. 

There are some **quirks and restrictions to be aware of while we strive towards 
that ideal**. Additionally, we've added backwards-compatibility features that make 
the transition from tape-based code to program-capture be a smooth one. In this 
document, we provide an overview of the constraints, "gotchas" to be aware of, and
features that will help get your existing tape-based code working with program-capture.

.. note::

>     #. From here onwards, :func:`~.pennylane.capture.enable` is assumed to be present 

Reviewer 3: Is the period at the beginning of every entry intended?

![SB1](https://github.com/user-attachments/assets/543a9ee2-0183-440f-a048-1f3cc5c36d00)

    within the scope of code examples, unless otherwise stated. This ensures that 
    program-capture is enabled.

>     #. Using program-capture requires that JAX be installed. Please consult the 
>     JAX documentation for `installation instructions <https://docs.jax.dev/en/latest/installation.html>`__.

Reviewer 3: ```suggestion
    #. Using program-capture requires that JAX is installed. Please consult the 
    JAX documentation for `installation instructions <https://docs.jax.dev/en/latest/installation.html>`__.
```

    
    #. Our short name for PennyLane code that is captured as jaxpr is *plxpr* (PennyLane-jaxpr).
    "Program-capture" and "plxpr" can be considered synonymous and interchangeable. 

Device compatibility
--------------------

Currently, ``default.qubit`` and ``lightning.qubit`` are the only devices compatible 
with program-capture.

Device wires 
~~~~~~~~~~~~

With program-capture enabled, both ``lightning.qubit`` and ``default.qubit`` require 
that ``wires`` be specified at device instantiation (this is in contrast to when 
program-capture is disabled, where automatic qubit management takes place internally
with ``default.qubit``).

.. code-block:: python 

    qml.capture.enable()

    @qml.qnode(qml.device('default.qubit'))
    def circuit():
        qml.Hadamard(0)
        return qml.state()

>>> circuit()
NotImplementedError: devices must specify wires for integration with plxpr capture.

> 

Reviewer: Is it worth following this example with one where it works? Perhaps use the `QNode.update` method?

```python
>>> circuit = circuit.update(device = qml.device('default.qubit', wires=1))
>>> circuit()
...
```
Could get the reader hyped? 😎

Author: Not sure what's happening 🥴

Reviewer: What's your `qml.about()` say?

Valid JAX data types 
--------------------

Because of the nature of creating and executing plxpr, it is **best practice to 
use JAX-compatible types whenever possible**, in particular for arguments to quantum 
functions and QNodes, and positional arguments in PennyLane gate operations. Examples of 
JAX-compatible types are ``jax.numpy`` arrays, regular NumPy arrays, dictionaries, standard
Python ``int``\ s and ``float``\ s, and anything else with a valid `Pytree <https://jax.readthedocs.io/en/latest/pytrees.html>` 
representation.

For example, ``list``\ s, ``range``\ s, and strings are not valid JAX types for 
the positional argument in :class:`~.pennylane.MultiRZ`, and will result in an error:

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device('default.qubit', wires=2)

    @qml.qnode(dev)
    def circuit():
        qml.MultiRZ([0.1, 0.2], wires=[0, 1])
        return qml.expval(qml.X(0))

>>> circuit()
TypeError: Value [0.1, 0.2] with type <class 'list'> is not a valid JAX type

> 

Reviewer: Again, maybe a quick blurb about replacing this with `jax.numpy.array([0.1, 0.2])`?

where the ``list`` is being indexed into, thereby retrieving a valid JAX type:

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device('default.qubit', wires=2)

    @qml.qnode(dev)
    def circuit(x):
        qml.RZ(x[0], wires=0)
        qml.RX(x[1], wires=1)
        return qml.expval(qml.X(0))

>>> circuit([0.1, 0.2])
Array(0., dtype=float32)

JAX-incompatible types, like Python ``range``\ s, are acceptable as keyword arguments:

.. code-block:: python 

    qml.capture.enable()
    
    dev = qml.device('default.qubit', wires=2)

    @qml.qnode(dev)
    def circuit(x, range_of_wires=None):
        for w in range_of_wires:
            qml.RZ(x[0], wires=w)
            qml.RX(x[1], wires=w)

        return qml.expval(qml.X(0))

>>> circuit([0.1, 0.2], range_of_wires=range(2))
Array(0., dtype=float32)

But, again, using JAX-compatible types wherever possible is recommended.

Positional arguments
~~~~~~~~~~~~~~~~~~~~

Positional arguments in PennyLane are flexible in that their variable names can 
instead be employed as keyword arguments (e.g., ``qml.RZ(0.1, wires=0)`` versus 
``qml.RZ(phi=0.1, wires=0)``). However, to ensure differentiability and, in general,

> compatilibty with program-capture enabled, such arguments must be kept as positional, 

Reviewer 3: ```suggestion
compatibility with program-capture enabled, such arguments must be kept as positional, 
```

                rxs.append(op)
            else:
                new_ops.append(op)
        
        operations = new_ops + rxs
        new_tape = tape.copy(operations=operations)
        return [new_tape], lambda res: res[0]

Decorating with just ``@shift_rx_to_end`` will not work, and will give a vague error:

.. code-block:: python 

    qml.capture.enable()

    @shift_rx_to_end
    @qml.qnode(qml.device("default.qubit", wires=1))
    def circuit():
        qml.RX(0.1, wires=0)
        qml.H(wires=0)
        return qml.state()

>>> print(qml.draw(circuit)())
...
NotImplementedError: 

A requirement for tape transforms to be compatible with program capture is to further 
decorate QNodes with the experimental :func:`~.pennylane.capture.expand_plxpr_transforms` 
decorator:

.. code-block:: python 

    qml.capture.enable()

    @qml.capture.expand_plxpr_transforms
    @shift_rx_to_end
    @qml.qnode(qml.device("default.qubit", wires=1))
    def circuit():
        qml.RX(0.1, wires=0)
        qml.H(wires=0)
        return qml.state()

>>> print(qml.draw(circuit)())
0: ──H──RX(0.10)─┤  State

Dynamic variables and transforms
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Some transforms in the :doc:`/code/qml_transforms` module have natively support program capture mode:

Author: ```suggestion
Some transforms in the :doc:`/code/qml_transforms` module have natively support program-capture:
```

#. :func:`~.pennylane.transforms.merge_rotations`
#. :func:`~.pennylane.transforms.single_qubit_fusion`
#. :func:`~.pennylane.transforms.unitary_to_rot`
#. :func:`~.pennylane.transforms.merge_amplitude_embedding`
#. :func:`~.pennylane.transforms.commute_controlled`
#. :func:`~.pennylane.transforms.decompose`
#. :func:`~.pennylane.map_wires`

> #. :func:`~.pennylane.transforms.cancel_inverses`

Reviewer: Where do we want to mention the limitations with the merging transforms where they cannot merge through higher-order primitives? For example, an `RX` before a for loop and any `R(X,Y,Z)` rotations inside of said for loop won't merge.

Author: I left a comment here to put this in 😁 https://github.com/PennyLaneAI/pennylane/pull/7062#discussion_r2006496987

Reviewer 3: Maybe it is ideal to mention that there are 'some limitations/differences' and show one example. I don't think there is a chance to mention all the specific differences for every transform in this document

See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerIntegerConversionError

while loops 
-----------

While loops written with :func:`~.pennylane.while_loop` cannot accept a ``lambda``
function:

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit():

        @qml.while_loop(lambda a: a > 3)
        def loop(a):
            a += 1
            return a

        a = 0
        loop(a)

        qml.RX(0, wires=0)
        return qml.state()

>>> circuit()
...
KeyError: <gast.gast.Lambda object at 0x136ff82b0>

As a workaround, use a regular Python function:

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    def func(x):
        return x > 3

    @qml.qnode(dev)
    def circuit():

        @qml.while_loop(func)
        def loop(a):
            a += 1
            return a

        a = 0
        loop(a)
        
        qml.RX(0, wires=0)
        return qml.state()

>>> circuit()

> Array([1.+0.j, 0.+0.j], dtype=complex64)

Reviewer: The work-around is even easier, you can just set the lambda to a callable variable (code below runs fine),
```python
qml.capture.enable()
dev = qml.device("default.qubit", wires=1)
@qml.qnode(dev)
def circuit():
    condition = lambda a: a > 3
    @qml.while_loop(condition)
    def loop(a):
        a += 1
        return a
    a = 0
    loop(a)
    qml.RX(0, wires=0)
    return qml.state()
```

Author: Oh nice! Thank you Andrija :)

Calculating operator matrices in QNodes
---------------------------------------

> The matrix of an operator cannot be computed with :func:`~.pennylane.matirx` within

Author: 🥴

a QNode, and will raise an error:

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit():
        mat = qml.matrix(qml.X(0))
        return qml.state()

>>> circuit()
...
TransformError: Input is not an Operator, tape, QNode, or quantum function

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit():
        mat = qml.matrix(qml.X)(0)
        return qml.state()

>>> circuit()
...
NotImplementedError: 

> Section title 

Author: This is here for copy-paste. Please remove before PR gets merged

> -------------

Reviewer 3: I guess all the remaining sharp-bits about decompositions, gradients, etc. will be added in different sections.

Author: @PietropaoloFrisoni which sharp bits do you have in mind re: decompositions and gradients? 🤔

Author: Thank you! Will add these :)

Author: Sorry guys do you mind adding a sharp bits entry for gradients? I need a from-source install of everything and I can't get the Catalyst installation working 🙏

Author: Also @PietropaoloFrisoni can you provide an example of the issue with operators that only have a `compute_decomposition` method? 😅

Author: https://github.com/PennyLaneAI/pennylane/pull/7062/files#r2019266477

Reviewer 3: Resolved on Slack (nothing to add here). Sorry for the late reply!

        qml.RX(0.1, wires=0)

        for i in range(4):
            qml.RX(0.1, wires=0)
            qml.RX(0.1, wires=0)

        qml.RX(0.1, wires=0)

        return qml.state()

    print(qml.capture.make_plxpr(circuit)())

The above example should result in a single ``RX`` gate with an angle of ``1.0``, 
but transforms are unable to transfer through the circuit in its entirety. Drawing
this circuit will result in an inaccurate circuit:

>>> print(qml.draw(circuit)())
0: ──RX(0.30)─┤  State

To illustrate what is actually happening internally, consider the plxpr representation 
of this program: 

>>> print(qml.capture.make_plxpr(circuit)())
{ ...
    qfunc_jaxpr={ lambda ; . let
        _:AbstractOperator() = RX[n_wires=1] 0.1 0
        for_loop[
          abstract_shapes_slice=slice(0, 0, None)
          args_slice=slice(0, None, None)
          consts_slice=slice(0, 0, None)
          jaxpr_body_fn={ lambda ; b:i32[]. let
              _:AbstractOperator() = RX[n_wires=1] 0.2 0
            in () }
        ] 0 4 1
        _:AbstractOperator() = RX[n_wires=1] 0.1 0
    ...
}

As one can see, the outer ``RX`` gates do not merge with those in the ``for`` loop, 

> nor does the transform "smartly" merge all 4 iterations from the ``for`` loop.

Author: ```suggestion
nor does the transform merge all 4 iterations from the ``for`` loop.
```

#. :func:`~.pennylane.transforms.cancel_inverses`

For transforms that do not natively work with program capture, they can continue to be used with certain limitations:

#. Transforms that return multiple tapes are not supported.
#. Transforms that return non-trivial post-processing functions are not supported.
#. Tape transforms will fail to execute if the transformed quantum function or QNode contains:

   #. ``qml.cond`` with dynamic parameters as predicates.
   #. ``qml.for_loop`` with dynamic parameters for ``start``, ``stop``, or ``step``.
   #. ``qml.while_loop``.

Here is an example with our toy ``shift_rx_to_end`` transform and a dynamic parameter
for ``stop`` in ``qml.for_loop``.

.. code-block:: python 

    qml.capture.enable()

    @qml.capture.expand_plxpr_transforms
    @shift_rx_to_end
    @qml.qnode(qml.device("default.qubit", wires=4))
    def circuit(stop):

        @qml.for_loop(0, stop, 1)
        def loop(i):
            qml.RX(0.1, wires=i)
            qml.H(wires=i)
        
        loop(stop)

        return qml.state()

>>> circuit(4)
TracerIntegerConversionError: The __index__() method was called on traced array with shape int32[].

> The error occurred while tracing the function wrapper at /Users/isaac/.virtualenvs/pl-latest/lib/python3.11/site-packages/pennylane/transforms/core/transform_dispatcher.py:41 for make_jaxpr. This concrete value was not available in Python because it depends on the value of the argument inner_args[0].

Author: ```suggestion
The error occurred while tracing the function wrapper at <path to environment>/site-packages/pennylane/transforms/core/transform_dispatcher.py:41 for make_jaxpr. This concrete value was not available in Python because it depends on the value of the argument inner_args[0].
```

.. role:: html(raw)
   :format: html

> 🔪 Program capture sharp bits

Author: @PietropaoloFrisoni lol good question! We want to move this down to the release news section, so we can re-evaluate when it moves there 😅

=============================

Program capture is a new feature of PennyLane that allows for compactly expressing 
details about hybrid workflows, including quantum operations, classical processing, 
structured control flow, and dynamicism.

This new feature unlocks better performance, more functional and dynamic workflows, 
and smoother integration with just-in-time compilation frameworks like 
`Catalyst <https://docs.pennylane.ai/projects/catalyst/en/stable/index.html>`__ 
(via the :func:`~.pennylane.qjit` decorator) and JAX-jit.

Internally, program capture is supported by representing hybrid programs via a new 
intermediate representation (IR) called ``plxpr``, rather than a quantum tape. The 
``plxpr`` IR is an adaptation of JAX's ``jaxpr`` IR.

Our vision with ``plxpr`` is to supplant the quantum tape as our default program 
representation, and to support *more* than just the core features of PennyLane.

> There are some **quirks and restrictions to be aware of while we strive towards 
> that ideal**. Additionally, we've added backwards-compatibility features that make 
> the transition from tape-based code to program capture be a smooth one. In this 

Reviewer 3: ```suggestion
There are some **quirks and restrictions to be aware of while we strive towards 
that ideal**. Additionally, we've added backward compatibility features that make 
the transition from tape-based code to program capture smooth. In this 
```

For this type of suggestion, I just report what Grammarly suggests in some cases, but I am not a native English speaker, so if something doesn't sound right to you, please ignore!

    dev = qml.device('default.qubit', wires=1)

    @qml.qnode(dev)
    def circuit(x):
        qml.RX(x, wires=0)
        m0 = qml.measure(0)
        return qml.state()

>>> circuit(0.1)
...
TransformError: Too many mid-circuit measurements for the specified number of wires with 'defer_measurements'.

Recall that the deferred measurements MCM method adds a temporary wire and represents 
the physical MCM as a controlled operation, deferring the measurement until the 
end of the circuit. By adding an additional wire to the device, the above circuit
executes as expected: 

>>> circuit = circuit.update(device = qml.device('default.qubit', wires=2))
>>> circuit(0.1)
Array([0.99875027+0.j        , 0.        +0.j        ,
       0.        +0.j        , 0.        -0.04997917j], dtype=complex64)

Valid JAX data types 
--------------------

Because of the nature of creating and executing plxpr, it is **best practice to 
use JAX-compatible types whenever possible**, in particular for arguments to quantum 
functions and QNodes, and positional arguments in PennyLane gate operations. Examples of 
JAX-compatible types are ``jax.numpy`` arrays, regular NumPy arrays, dictionaries, standard

> Python ``int``\ s and ``float``\ s, and anything else with a valid `Pytree <https://jax.readthedocs.io/en/latest/pytrees.html>` 

Author: Something wrong with how the Pytree link is rendering

Author: ```suggestion
Python ``int``\ s and ``float``\ s, and anything else with a valid `Pytree <https://jax.readthedocs.io/en/latest/pytrees.html>`__
```

    dev = qml.device('default.qubit', wires=2)

    @qml.qnode(dev)
    def circuit(x):
        qml.RZ(x[0], wires=0)
        qml.RX(x[1], wires=1)
        return qml.expval(qml.X(0))

>>> circuit([0.1, 0.2])
Array(0., dtype=float32)

Keyword arguments
~~~~~~~~~~~~~~~~~

JAX-incompatible types, like Python ``range``\ s, are acceptable as **keyword arguments**:

.. code-block:: python 

    qml.capture.enable()
    
    dev = qml.device('default.qubit', wires=2)

    @qml.qnode(dev)
    def circuit(x, range_of_wires=None):
        for w in range_of_wires:
            qml.RZ(x[0], wires=w)
            qml.RX(x[1], wires=w)

        return qml.expval(qml.X(0))

>>> circuit([0.1, 0.2], range_of_wires=range(2))
Array(0., dtype=float32)

But, again, using JAX-compatible types wherever possible is recommended.

Positional arguments
~~~~~~~~~~~~~~~~~~~~

Positional arguments in PennyLane are flexible in that their variable names can 
instead be employed as keyword arguments (e.g., ``qml.RZ(0.1, wires=0)`` versus 
``qml.RZ(phi=0.1, wires=0)``). However, to ensure differentiability and, in general,
compatibility with program capture enabled, such arguments must be kept as positional, 

> regardless of if they're provided as an acceptable JAX type. 

Reviewer 3: ```suggestion
regardless of whether they're provided as an acceptable JAX type. 
```

Using program capture with Catalyst
-----------------------------------

To use the program capture feature with Catalyst, the ``qml.capture.enable()`` toggle
is not required. Instead, when decorating a workflow with :func:`~.pennylane.qjit`, 
add the ``experimental_capture=True`` flag:

.. code-block:: python

    dev = qml.device('lightning.qubit', wires=1)

    @qml.qjit(experimental_capture=True)
    @qml.qnode(dev)
    def circuit():
        qml.RX(0.1, wires=0)
        return qml.state()

>>> circuit()
Array([0.99875026+0.j        , 0.        -0.04997917j], dtype=complex128)

Transforms
----------

One of the core features of PennyLane is modularity, which has allowed users to 
transform QNodes in a NumPy-like way and to create their own transforms with ease. 
Your favourite transforms will still work with program capture enabled (including
custom transforms), but decorating QNodes with just ``@transform_name`` **will not 
work** and will give a vague error. Additionally, decorating QNodes with the experimental 
:func:`~.pennylane.capture.expand_plxpr_transforms` decorator is required.

Consider the following toy example, which shows a tape-based transform that shifts 

> all :class:`~.pennylane.RX`` gates to the end of a circuit.

Reviewer 3: ```suggestion
all :class:`~.pennylane.RX` gates to the end of a circuit.
```

.. code-block:: python 

    qml.capture.enable()

    @qml.transform
    def shift_rx_to_end(tape):
        """Transform that moves all RX gates to the end of the operations list."""
        new_ops, rxs = [], []

        for op in tape.operations:
            if isinstance(op, qml.RX):
                rxs.append(op)
            else:
                new_ops.append(op)
        
        operations = new_ops + rxs
        new_tape = tape.copy(operations=operations)
        return [new_tape], lambda res: res[0]

Decorating with just ``@shift_rx_to_end`` will not work, and will give a vague error:

.. code-block:: python 

    qml.capture.enable()

    @shift_rx_to_end
    @qml.qnode(qml.device("default.qubit", wires=1))
    def circuit():
        qml.RX(0.1, wires=0)
        qml.H(wires=0)
        return qml.state()

>>> print(qml.draw(circuit)())
...
NotImplementedError: 

> A requirement for tape transforms to be compatible with program capture is to further 

Reviewer 3: ```suggestion
A requirement for transforms to be compatible with program capture is to further 
```

This is true even for transforms like `merge_rotations` with a capture-native implementation which does not use tapes, right?

Author: Thanks Pietro!

decorate QNodes with the experimental :func:`~.pennylane.capture.expand_plxpr_transforms` 
decorator:

.. code-block:: python 

    qml.capture.enable()

    @qml.capture.expand_plxpr_transforms
    @shift_rx_to_end
    @qml.qnode(qml.device("default.qubit", wires=1))
    def circuit():
        qml.RX(0.1, wires=0)
        qml.H(wires=0)
        return qml.state()

>>> print(qml.draw(circuit)())
0: ──H──RX(0.10)─┤  State

Control flow and transforms
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Transforms do not apply "through" control flow when capture is enabled. An example

> is best to demonstrate this behaviour:

Reviewer: This is fine, but its not just control flow - it's any higher order primitive (mid measures, grad, jacobian, for, while, conditionals) 😄

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device('default.qubit', wires=1)

    @qml.capture.expand_plxpr_transforms
    @qml.transforms.merge_rotations
    @qml.qnode(dev)
    def circuit():
        qml.RX(0.1, wires=0)

>         for i in range(4):

Reviewer 3: ```suggestion
        for _ in range(4):
```

            qml.RX(0.1, wires=0)
            qml.RX(0.1, wires=0)

        qml.RX(0.1, wires=0)

        return qml.state()

> 
>     print(qml.capture.make_plxpr(circuit)())

Author: ```suggestion
```

The above example should result in a single ``RX`` gate with an angle of ``1.0``, 
but transforms are unable to transfer through the circuit in its entirety. Drawing
this circuit will result in an inaccurate circuit:

>>> print(qml.draw(circuit)())

> 0: ──RX(0.30)─┤  State

Author: This example seems a little out-of-place?

Reviewer 3: Maybe to be on the safe side, we can just say that program visualisation is not fully compatible with program capture? 
As you rightly said at the beginning, PennyLane is largely built around the concept of tape, while now we have to deal with `jaxpr`, over which we have much less control :(

> 0: ──RX(0.30)─┤  State

Reviewer 3: ```suggestion
0: ──RX(0.20)─┤
```

This is my output, which is different, although still incorrect. I also tried other cases, but I can't find a clear logic in what is printed, so it seems to me that visualisation is simply incompatible with program capture

    def circuit(stop):

        @qml.for_loop(0, stop, 1)
        def loop(i):
            qml.RX(0.1, wires=i)
            qml.H(wires=i)
        
        loop(stop)

        return qml.state()

>>> circuit(4)
TracerIntegerConversionError: The __index__() method was called on traced array with shape int32[].
The error occurred while tracing the function wrapper at /Users/isaac/.virtualenvs/pl-latest/lib/python3.11/site-packages/pennylane/transforms/core/transform_dispatcher.py:41 for make_jaxpr. This concrete value was not available in Python because it depends on the value of the argument inner_args[0].
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerIntegerConversionError

Dynamic shapes
--------------

Creating and manipulating dynamically shaped objects within a quantum function or 
QNode when capture is enabled is supported with 
`JAX's experimental dynamic shapes support <https://docs.jax.dev/en/latest/notebooks/Common_Gotchas_in_JAX.html#dynamic-shapes>`__. 
Given the experimental nature of this feature, PennyLane's dynamic shapes support 
is at best a subset of what is possible with purely classical programs using JAX. 

> 

Reviewer: Do we want an example with dynamic shapes?

Reviewer 3: Good point, or at least explain why we don't want to provide it if that is the case.

Parameter broadcasting and vmap
-------------------------------

Parameter-broadcasting is generally not compatible with program capture. There are 
cases that magically work, but one shouldn't extrapolate beyond those particular 
cases.

Instead, it is best practice to `use jax.vmap <https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html>`__:

.. code-block:: python 

>     qml.capture.enable()

Reviewer 3: ```suggestion
    import jax

    qml.capture.enable()
```

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit():

        func = lambda x: x > 3

        @qml.while_loop(func)
        def loop(a):
            a += 1
            return a

        a = 0
        loop(a)
        
        qml.RX(0, wires=0)
        return qml.state()

>>> circuit()
Array([1.+0.j, 0.+0.j], dtype=complex64)

or use a regular Python function,

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    def func(x):
        return x > 3

    @qml.qnode(dev)
    def circuit():

        @qml.while_loop(func)
        def loop(a):
            a += 1
            return a

        a = 0
        loop(a)
        
        qml.RX(0, wires=0)
        return qml.state()

>>> circuit()
Array([1.+0.j, 0.+0.j], dtype=complex64)

Calculating operator matrices in QNodes
---------------------------------------

The matrix of an operator cannot be computed with :func:`~.pennylane.matrix` within
a QNode, and will raise an error:

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit():

>         mat = qml.matrix(qml.X(0))

Reviewer 3: ```suggestion
        qml.matrix(qml.X(0))
```

Is this equivalent? It seems to provide the same error:

```
TransformError: Input is not an Operator, tape, QNode, or quantum function
```

        return qml.state()

>>> circuit()
...
TransformError: Input is not an Operator, tape, QNode, or quantum function

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit():

>         mat = qml.matrix(qml.X)(0)

Reviewer 3: same here

    def circuit(x):
        qml.RX(x, wires=0)
        return qml.expval(qml.Z(0))

>>> x = jnp.array([0.1, 0.2, 0.3])
>>> vmap_circuit = jax.vmap(circuit)
>>> vmap_circuit(x)
Array([0.9950042 , 0.9800666 , 0.95533645], dtype=float32)

More information for using ``jax.vmap`` can be found in the 
`JAX documentation <https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html#jax.vmap>`__.

Decompositions
--------------

With program capture enabled, operators present in a circuit with control flow 
in their ``compute_decomposition`` method and who also do not have a ``compute_qfunc_decomposition``
method will cause an error. The ``StronglyEntanglingLayers`` template is an example
of such an operator:

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device('default.qubit', wires=4)

    @qml.qnode(dev)
    def circuit(parameters):
        qml.StronglyEntanglingLayers(weights=parameters, wires=range(4))
        return qml.expval(qml.Z(0))

    shape = qml.StronglyEntanglingLayers.shape(n_layers=2, n_wires=4)
    weights = np.random.random(size=shape)

>>> circuit(weights)
TODO

This is not an issue for operators with a ``compute_qfunc_decomposition`` method, 

> TODO.

Author: cc @PietropaoloFrisoni

Reviewer 3: Resolved on Slack :+1:

Positional arguments
~~~~~~~~~~~~~~~~~~~~

Positional arguments in PennyLane are flexible in that their variable names can 
instead be employed as keyword arguments (e.g., ``qml.RZ(0.1, wires=0)`` versus 
``qml.RZ(phi=0.1, wires=0)``). However, to ensure differentiability and, in general,
compatibility with program capture enabled, such arguments must be kept as positional, 
regardless of whether they're provided as an acceptable JAX type. 

For instance, consider this example with ``qml.RZ``:

.. code-block:: python 

    import jax.numpy as jnp

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit(angle):
        qml.RX(phi=angle, wires=0)
        return qml.expval(qml.Z(0))

>>> angle = jnp.array(0.1)
>>> circuit(angle)
...
UnexpectedTracerError: Encountered an unexpected tracer. A function transformed by JAX had a side effect, allowing for a reference to an intermediate value with type float32[] wrapped in a DynamicJaxprTracer to escape the scope of the transformation.
...

Even though the value for ``phi`` in ``qml.RZ`` is given as a valid JAX type, the 

> fact that it was provided as a keyword argument results in an error.

Reviewer: @josh146 @isaacdevlugt, in our current implementation we treat positional arguments as traceable and keyword arguments as "metadata". So, in this case if you try to specify `phi` as a kwarg, it errors out since it should be a traceable argument. 😄

> fact that it was provided as a keyword argument results in an error.

Reviewer 2: got it -- does this deviate from how JAX works by itself? If so, I'm curious why we chose to do this.

> fact that it was provided as a keyword argument results in an error.

Author: @andrijapau just following up 😁

        qml.RX(angle, wires=0)
        return qml.expval(qml.Z(0))

>>> angle = jnp.array(0.1)
>>> circuit(angle)
Array(0.9950042, dtype=float32)

Using program capture with Catalyst
-----------------------------------

To use the program capture feature with Catalyst, the ``qml.capture.enable()`` toggle
is not required. Instead, when decorating a workflow with :func:`~.pennylane.qjit`, 
add the ``experimental_capture=True`` flag:

.. code-block:: python

    dev = qml.device('lightning.qubit', wires=1)

    @qml.qjit(experimental_capture=True)
    @qml.qnode(dev)
    def circuit():
        qml.RX(0.1, wires=0)
        return qml.state()

>>> circuit()
Array([0.99875026+0.j        , 0.        -0.04997917j], dtype=complex128)

Transforms
----------

One of the core features of PennyLane is modularity, which has allowed users to 
transform QNodes in a NumPy-like way and to create their own transforms with ease. 
Your favourite transforms will still work with program capture enabled (including
custom transforms), but decorating QNodes with just ``@transform_name`` **will not 
work** and will give a vague error. Additionally, decorating QNodes with the experimental 

> :func:`~.pennylane.capture.expand_plxpr_transforms` decorator is required.

Author: This might not be a sharp bit! Waiting to see if this gets merged https://github.com/PennyLaneAI/pennylane/pull/7199

is at best a subset of what is possible with purely classical programs using JAX. 

To use JAX's experimental dynamic shapes support, you must add the following toggle 
to the top level of your program: 

.. code-block:: python

    jax.config.update("jax_dynamic_shapes", True)

Parameter broadcasting and vmap
-------------------------------

Parameter-broadcasting is generally not compatible with program capture. There are 
cases that magically work, but one shouldn't extrapolate beyond those particular 
cases.

Instead, it is best practice to `use jax.vmap <https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html>`__:

.. code-block:: python 

    import jax

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=1)

    @qml.qnode(dev)
    def circuit(x):
        qml.RX(x, wires=0)
        return qml.expval(qml.Z(0))

>>> x = jnp.array([0.1, 0.2, 0.3])
>>> vmap_circuit = jax.vmap(circuit)
>>> vmap_circuit(x)
Array([0.9950042 , 0.9800666 , 0.95533645], dtype=float32)

More information for using ``jax.vmap`` can be found in the 
`JAX documentation <https://docs.jax.dev/en/latest/_autosummary/jax.vmap.html#jax.vmap>`__.

> Decompositions

Author: @PietropaoloFrisoni let me know what you think!

--------------

> With program capture enabled, operators present in a circuit with control flow
> in their ``compute_decomposition`` method and who also do not have a ``compute_qfunc_decomposition``
> method may raise an error if you apply the decompose transform. This can happen, 
> for example, if the predicate of a conditional control flow operation specified 
> in a decomposition is a traced argument. 
> 
> For example, the :class:`~.pennylane.RandomLayers` template does not have a ``compute_qfunc_decomposition``
> method, and its ``compute_decomposition`` method contains an ``if`` statement that 
> predicates on a traced argument:
> 

Reviewer 3: ```suggestion
With program capture enabled, operators used in circuits may raise an error when the ``decompose`` transform is applied.

This can happen if the operator:
- defines a ``compute_decomposition`` method that contains control flow (e.g., ``if`` statements),
- does not define a ``compute_qfunc_decomposition`` method, and
- receives a traced argument as part of the control flow condition.

For example, the :class:`~.pennylane.RandomLayers` template does not implement a ``compute_qfunc_decomposition`` method. Its ``compute_decomposition`` method includes an ``if`` statement where the condition depends on the ``ratio_imprim`` argument. If ``ratio_imprim`` is passed as a traced JAX value, an error occurs:

```

.. code-block:: python 

    qml.capture.enable()

    dev = qml.device("default.qubit", wires=2)

    @qml.capture.expand_plxpr_transforms
    @qml.transforms.decompose
    @qml.qnode(dev)
    def circuit(weights, arg):
        qml.RandomLayers(weights, wires=[0, 1], ratio_imprim=arg)
        return qml.expval(qml.Z(0))

    weights=jax.numpy.array([[0.1, -2.1, 1.4]])
    arg = 0.5

    circuit(weights, arg)

>>> weights=jax.numpy.array([[0.1, -2.1, 1.4]])
>>> arg = 0.5
>>> circuit(weights, arg)
...
The error occurred while tracing the function eval at pennylane/transforms/decompose.py:243 for jit. This value became a tracer due to JAX operations on these lines:
  operation a:bool[] = lt b c
    from line pennylane/templates/layers/random.py:245:19 (RandomLayers.compute_decomposition)
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerBoolConversionError

> 

Reviewer 3: If you think that it is the case, we can also add something as the following (and maybe clarify that in the future every operator will have a `compute_qfunc_decomposition`):

```suggestion

In this case, we can pass ``ratio_imprim`` as a regular (non-traced) constant:

-------------------------------------------------------
dev = qml.device("default.qubit", wires=2)

@qml.capture.expand_plxpr_transforms
@qml.transforms.decompose
@qml.qnode(dev)
def circuit1(weights):
    qml.RandomLayers(weights, wires=[0, 1], ratio_imprim=0.5)
    return qml.expval(qml.Z(0))

>> circuit1(jax.numpy.array([[0.1, -2.1, 1.4]]))
Array(0.99500424, dtype=float32)
-------------------------------------------------------

Currently, the operators that define a ``compute_qfunc_decomposition`` are:
- ``StronglyEntanglingLayers``
- ``GroverOperator``
- ``QFT``

```

 I don't know how to fix indentation and rendering in a GitHub suggestion : )
 
 There would be more caveats around this, but I don't think we should go into this level of detail, so I hope this is enough. What do you think?
